### DISCUSSION TITLE
Port Server and socat

### DISCUSSION THREAD

**Discussion by Wolfillo** [metrafonic/asyncio-vedirect-mqtt#6] [Q&A]
Hi again! 

I'm trying to do a new deployment of the MPPT 75|15 using a Teltonika 4G router with a serial port server to send the ve.direct data over IP. I succesfully attached the ve.direct serial port to the serial port input of the router. Then I created a port server in the router. As the router has a CLI, I can see that the router is recieving the data over the rs232 port:
![2024-08-06 12_18_09-TRB245 - Teltonika Networks](https://github.com/user-attachments/assets/755bc1a0-77be-4645-a415-be0ac20284a8)

Then, in a Linux Mint VM (in the same network) I open a socat to create a virtual serial port:
`sudo socat PTY,link=/dev/ttyV0,echo=0,raw TCP:192.168.141.254:10991`

When I cat the virtual port (ttyV0) I get the same output as in the router so I assume that's ok. But when I run the ve-mqtt, it doesn't publish any data of the sensors. I don't now if I'm missing something but I can't figure out what...

Thank you in advance.

**Comment by metrafonic** [comment]
Hi @Wolfillo !
I like that you are able to use a teltonika router connected directly to the SmartSolar device. I'm a big fan of the TRB series they make. Can I ask which model you are using? I have a couple of TRB140s, but unfortunately they do not have any exposed RS232 Ports, so I could never get it to work on them directly.

Regarding the text output, it looks like maybe the baud rate is incorrect. The text output should show human readable values. Are you able to change the baud rate?

The `tests/smartsolar_1.39.dump` file shows how this output should look like.

**Reply by Wolfillo** [reply]
Hello @metrafonic !

It's a TRB245 4G router/moden wich has an embbeded rs232/rs485 port server with plenty of protocols and modes to use. In my case I wanted to get rid of the RBPi as I already had this device. I think I can even make your code to run in the RutOS environement of the teltonika but first I wanted to make sure it works the way I intended to.

![image](https://github.com/user-attachments/assets/5e6dea96-2a51-4b51-a945-b1579cc5cd15)

This is the rs232 configuration, I think everything is correct. With this, I open a port server and then I socat that IP/Port to a ttyV0.

But the issue seems to come from the rs232 readings I'm afraid. I'm going to try with a raspi and see what it sees from it's serial.

Regards!

**Reply by Wolfillo** [reply]
Ok, I can see something is happening with the serial port in the Teltonika...I just read the serial output from the victron, just using a rusty serial to usb I had laying arround and a PuTTy:
![image](https://github.com/user-attachments/assets/8b86f61d-7add-43ec-8d34-3d05600eebcd)
I'm going to play arround with the TRB serial settings and see what happens.

Thank you

**Reply by metrafonic** [reply]
Did you end up getting it to work straight from the teltonika?

**Reply by Wolfillo** [reply]
Hi!

Not yet...I'm still in contact with Teltonika to solve the issue but...no chances yet.

**Reply by Wolfillo** [reply]
Greetings! I have some good news.

I found myself a little bit dumb with the answer they gave me... It was the serial voltage levels. The MPPT manages 5V TTL serial but the TRB245 is pure RS232 so I had to put a MAX3232 chip converter to adapt the serial signal.

Well, to my credit I have to say there's no information about this in the TRB245 manual and the device's port is just a bare quick connector, not a DB9 or something that would remind you it is not TTL. Futhermore, the construction of the device and the software could make you think it's a raspi with network capabilities, which could be true since it runs RutOS based in Linux.

And speaking of that I'm going to try to install the vedirect-mqtt inside the TRB since it supports Python... wish me luck.

Thank you.
